src/js.c

```diff
@@ -2827,6 +2903,220 @@ js_PullSubscribe(natsSubscription **sub, jsCtx *js, const char *subject, const c
     return NATS_UPDATE_ERR_STACK(s);
 }
 
+// sub->mu must NOT be held.
+natsStatus
+js_maybeFetchMore(natsSubscription *sub, jsFetch *fetch)
+{
+    jsFetchRequest req;
+    if (fetch->nextf == NULL)
+        return NATS_OK;
+
+    // Prepare the next fetch request
+    if (!fetch->nextf(&req, sub, fetch->nextClosure))
+        return NATS_OK;
+
+    // These are not changeable by the callback, only Batch and MaxBytes can be updated.
+    int64_t now = nats_Now();
+    req.Heartbeat = fetch->lifetime.Heartbeat;
+    req.Expires = fetch->lifetime.Expires - (now - fetch->startTimeMilli) * 10*1000*1000;
+    req.NoWait = fetch->lifetime.NoWait;
+
+    char buffer[128];
+    natsBuffer buf;
+    natsBuf_InitWithBackend(&buf, buffer, 0, sizeof(buffer));
+
+    natsSub_Lock(sub);
+
+    jsSub *jsi = sub->jsi;
+    jsi->inFetch = true;
+    jsi->fetchID++;
+    snprintf(fetch->replySubject, sizeof(fetch->replySubject), "%.*s%" PRIu64,
+             (int)strlen(sub->subject) - 1, sub->subject, // exclude the last '*'
+             jsi->fetchID);
+
+    natsStatus s = _sendPullRequest(sub->conn, jsi->nxtMsgSubj, fetch->replySubject, &buf, &req);
+    if (s == NATS_OK)
+    {
+        fetch->requestedMsgs += req.Batch;
+    }
+    natsSub_Unlock(sub);
+
+    natsBuf_Destroy(&buf);
+    return NATS_UPDATE_ERR_STACK(s);
+}
+
+static bool
+_autoNextFetchRequest(jsFetchRequest *req, natsSubscription *sub, void *closure)
+{
+    jsFetch *fetch                  = (jsFetch *)closure;
+    int     remainingUnrequested    = 0;
+    int64_t remainingBytes          = 0;
+    int     want                    = 0;
+    bool    maybeMore               = true;
+
+    natsSub_Lock(sub);
+
+    int isAhead = fetch->requestedMsgs - fetch->deliveredMsgs;
+    int wantAhead = fetch->keepAhead;
+    if (isAhead > wantAhead)
+        maybeMore = false;
+
+    if (maybeMore)
+    {
+        // fetch->lifetime.Batch is always > 0
+        remainingUnrequested = fetch->lifetime.Batch - fetch->requestedMsgs;
+        if (remainingUnrequested <= 0)
+            maybeMore = false;
+    }
+
+    if (maybeMore && (fetch->lifetime.MaxBytes > 0))
+    {
+        remainingBytes = fetch->lifetime.MaxBytes - fetch->receivedBytes;
+        if (remainingBytes <= 0)
+            maybeMore = false;
+    }
+
+    if (maybeMore)
+    {
+        want = remainingUnrequested;
+        if (want > fetch->fetchSize)
+            want = fetch->fetchSize;
+
+        maybeMore = (want > 0);
+    }
+
+    natsSub_Unlock(sub);
+
+    if (!maybeMore)
+        return false;
+
+    *req = fetch->lifetime; // copy bytes, reading immutable data
+    req->Batch = want;
+    // FIXME discuss in PR - this seems wrong, we don't know how many bytes we will have
```

Comment:
Note: still unresolved, and still not quite sure how

Reply:
@kozlovic I am at a loss on how to properly resolve this, short of erroring on the side of over-asking and enforcing the limit in the client, which I do not like.

Reply:
I don't know either.

Reply:
No longer allow `MaxBytes` and `KeepAhead` at the same time; With that, the math for the next request's MaxBytes should be accurate.